Make import path specs consistent and fix error when running tests that use jsx/tsx

grubbins opened the pull request:
When jest tests import a `jsx`/`tsx` module, they fail with this obscure error:
```
 FAIL  src/react_fragment.test.tsx
  ● Test suite failed to run

    TypeError: React.createContext is not a function

      at Object.<anonymous> (node_modules/@emotion/react/dist/emotion-element-e89f38a3.cjs.dev.js:41:48)
      at Object.<anonymous> (node_modules/@emotion/react/dist/emotion-react.cjs.dev.js:7:22)

```

The explanation is:
* The use of React fragments causes `emotion` to be loaded
* `emotion` also imports React with `require('react')`
* This causes an ambiguity for the jest loader. The `moduleDirectories` includes both `"node_modules"` and `""`, leading to ambiguity between the actual React module and `emotion/react`. The result of the import ends up being `undefined` which later leads to the error shown.

A similar description of this issue in another project can be seen [here](https://github.com/next-page-tester/next-page-tester/issues/202#issuecomment-813092834).

This is resolved by removing `""` from `moduleDirectories` which seems to be a general recommendation anyway. I note that `mattermost-plugin-demo` does not have this configuration.

The only thing that is lost is the ability to `import ... from 'mymodule'` from a test file, but this can just be replaced by `import ... from './mymodule'`.

I added a test that reproduces the error. (I also confirmed the behaviour is the same whether or not we remove `emotion/core`, as I did in another PR). The test works after the jest config change.

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

fmartingr reviewed (commented):
Took a quick glance at this, not very familiar with the ecosystem.

hmhealey commented on webapp/package.json:
```diff
@@ -97,7 +97,6 @@
       "^bundle-loader\\?lazy\\!(.*)$": "$1"
     },
     "moduleDirectories": [
-      "",
```
That would match the web app as well since it currently has this set to `["", "node_modules"]`

fmartingr replied:
I have no strong opinions one way or the other. So I am tentatively approving this change, if someone else want to oppose the change please add your own review.

grubbins commented:
Modified to configure only `src` and `node_modules` as the search paths, for webpack, jest, and ts. I believe they will all be consistent now, and `import manifest from 'manifest'` works for all three.

The `@` alias for the `src` dir (which was configured for webpack and ts, but not jest) has thus been removed. The `webapp/` path itself was also removed from the webpack search paths. I believe this is now consistent with the mattermost project's `webapp/channels` if I am reading rightly.

@hmhealey you mentioned eslint config as well - but I can't see any specific path config for eslint, and it's also not throwing any errors. I think it is probably defaulting to using the typescript import path spec.

hmhealey reviewed (approved):
Sorry for the delay. Looks good to me!
